cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Constructs the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
+
+    @cached_property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self._range.left
+
+    @cached_property
+    def direction(self) -> str:
+        """``'to'`` if indexes are ascending, ``'downto'`` otherwise."""
+        return self._range.direction
+
+    @cached_property
+    def right(self) -> int:
+        """Rightmost index of the array."""
+        return self._range.right
+
+    @cached_property
+    def length(self):
+        """Length of the array."""
+        return self._range.length
+
+    @cached_property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._value)
+
+    def __reversed__(self) -> Iterator[Any]:
+        return reversed(self._value)
+
+    def __contains__(self, item: Any) -> bool:
+        return item in self._value
+
+    def __eq__(self, other: Any) -> bool:
+        try:
+            return all(a == b for a, b in zip_longest(self, other, fillvalue=object()))
+        except TypeError:
+            return NotImplemented
+
+    @overload
+    def __getitem__(self, item: int) -> Any:
+        pass  # pragma: no cover
+
+    @overload
+    def __getitem__(self, item: slice) -> "Array":
+        pass  # pragma: no cover
+
+    def __getitem__(self, item):
+        if isinstance(item, int):
+            idx = self._translate_index(item)
+            return self._value[idx]
+        elif isinstance(item, slice):
+            start = item.start or self.left
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
```

Comment:
A potential candidate for printing the directions or slice and array bounds. I have miscalculated slices in the past...

Reply:
Added info. It now reads.

```python
>>> from cocotb.types import Array, Range
>>> a = Array("1234")
>>> a[3:0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kaleb/dev/cocotb/cocotb/types/array.py", line 246, in __getitem__
    raise IndexError(
IndexError: slice [3:0] direction does not match array direction [0:3]
```